File: src/Triangle.java
import java.util.List;

public class Triangle {
    public int minimumTotal(List<List<Integer>> triangle) {
        for (int i = triangle.size() - 1; i > 0; i--) {
            List<Integer> current = triangle.get(i);
            List<Integer> prev = triangle.get(i - 1);
            for (int j = 0; j < prev.size(); j++) {
                int t = prev.remove(j);
                prev.add(j, t + Math.min(current.get(i), current.get(i + 1)));
            }
        }
        return triangle.get(0).get(0);
    }
}

File: Python/Triangle.py
from typing import List


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        n = len(triangle)
        dp = triangle[0]
        for i in range(2, 1 + n):
            new_dp = []
            for j in range(i):
                if j == 0:
                    new_dp.append(triangle[i - 1][j] + dp[0])
                elif j == i - 1:
                    new_dp.append(triangle[i - 1][j] + dp[-1])
                else:
                    new_dp.append(triangle[i - 1][j] + min(dp[j - 1], dp[j]))
            dp = new_dp[:]
        return min(dp)
